src/react/DevicesPage.jsx

```diff
@@ -35,126 +36,150 @@ import SingleCardPage from './ui/SingleCardPage';
 import { useAlerts } from './AlertManager';
 
 const DEVICES_PER_PAGE = 20;
+const DEVICES_PER_REQUEST = 100;
 const MAX_SHOWN_PAGES = 10;
 
+const matchFilters = (device, filters) => {
+  const {
+    deviceId = '',
+    showConnected = true,
+    showDisconnected = true,
+    showNeverConnected = true,
+  } = filters;
+
+  if (!showConnected && device.connected) {
+    return false;
+  }
+  if (!showDisconnected && !device.connected && device.lastConnection) {
+    return false;
+  }
+  if (!showNeverConnected && !device.connected && !device.lastConnection) {
+    return false;
+  }
+  if (deviceId === '') {
+    return true;
+  }
+
+  const aliases = Array.from(device.aliases.values());
+  return (
+    device.id.includes(deviceId) || aliases.filter((alias) => alias.includes(deviceId)).length > 0
+  );
+};
+
 export default ({ astarte, history }) => {
   const [phase, setPhase] = useState('loading');
-  const [totalDevices, setTotalDevices] = useState(null);
   const [activePage, setActivePage] = useState(0);
-  const [maxPage, setMaxPage] = useState(0);
-  const [cachedPages, setCachedPages] = useState([]);
-  const pageAlerts = useAlerts();
+  const [deviceList, setDeviceList] = useState([]);
+  const [requestToken, setRequestToken] = useState(false);
+  const [showSidebar, setShowSidebar] = useState(true);
+  const [filters, setFilters] = useState({});
 
-  const handleDevicesRequest = (pageIndex, response) => {
-    const token = new URLSearchParams(response.links.next).get('from_token');
-    const deviceList = response.data.map((value) => Device.fromObject(value));
-    cachedPages[pageIndex] = {
-      devices: deviceList,
-      token,
-    };
-    setCachedPages([...cachedPages]);
-    if (pageIndex === activePage) {
-      setPhase('ok');
-    }
-    const pagesToCache = activePage + MAX_SHOWN_PAGES + 1;
-    if (pageIndex < pagesToCache) {
-      // eslint-disable-next-line no-use-before-define
-      cachePage(pageIndex + 1);
+  const pageAlerts = useAlerts();
+  const pagedDevices = useMemo(() => {
+    const pageList = [];
+    const devices = deviceList.filter((device) => matchFilters(device, filters));
+    while (devices.length > 0) {
+      pageList.push(devices.splice(0, DEVICES_PER_PAGE));
     }
-    return null;
-  };
+    return pageList;
```

Comment:
This could be greatly simplified with Lodash: `const pageList = _.chunk(devices, DEVICES_PER_PAGE);`